File: nesoy.kwon/src/main/java/homework/chapter_7/BaseEntity.java
package homework.chapter_7;

import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;

import java.time.LocalDateTime;

/**
 * 아래의 형태처럼 들어감.
 *     create table Item (
 *         basePrice bigint not null,
 *         createdAt timestamp(6),
 *         id bigint not null,
 *         updatedAt timestamp(6),
 *         name varchar(255) not null,
 *         primary key (id)
 *     )
 */
@MappedSuperclass
public abstract class BaseEntity {
	protected LocalDateTime createdAt;
	protected LocalDateTime updatedAt;

	/**
	 * 저장하기 전에 진행한다.
	 */
	@PrePersist
	protected void prePersist() {
		createdAt = LocalDateTime.now();
		updatedAt = LocalDateTime.now();
	}

	/**
	 * 업데이트하기 전에 진행한다.
	 */
	@PreUpdate
	protected void preUpdate() {
		updatedAt = LocalDateTime.now();
	}
}

File: gracie.lee/src/main/kotlin/homework/chapter_7/BaseEntity.kt
package homework.chapter_7

import jakarta.persistence.*
import java.time.LocalDateTime

@MappedSuperclass //  // 테이블 없음, 공통 필드만 자식에게 상속
abstract class BaseEntity(
    protected var createdAt: LocalDateTime? = null,
    protected var updatedAt: LocalDateTime? = null
) {
    //  INSERT 전
    @PrePersist
    fun prePersist() {
        createdAt = LocalDateTime.now()
        updatedAt = LocalDateTime.now()
    }

    //  UPDATE 전
    @PreUpdate
    fun preUpdate() {
        updatedAt = LocalDateTime.now()
    }
}
